Gems/Atom/Feature/Common/Assets/Materials/Types/Debug/DebugVertexStreams.azsl

```diff
@@ -0,0 +1,134 @@
+    float m_bitangentSign : UV8;
+
+    // Extended fields (only referenced in this azsl file)...
+    float2 m_uv[UvSetCount] : UV1;
+};
+
+VSOutput MainVS(VSInput IN)
+{
+    VSOutput OUT;    
+    OUT.m_worldPosition = mul(ObjectSrg::GetWorldMatrix(), float4(IN.m_position, 1.0)).xyz;
+    OUT.m_position = mul(ViewSrg::m_viewProjectionMatrix, float4(OUT.m_worldPosition, 1.0));
+
+    // Only UV0 is supported
+    OUT.m_uv[0] = IN.m_uv0;
+    OUT.m_uv[1] = IN.m_uv0;    
+    
+    float4x4 objectToWorld = ObjectSrg::GetWorldMatrix();
+    float3x3 objectToWorldIT = ObjectSrg::GetWorldMatrixInverseTranspose();
+
+    ConstructTBN(IN.m_normal, IN.m_tangent, IN.m_bitangent, objectToWorld, objectToWorldIT, OUT.m_normal, OUT.m_tangent.xyz, OUT.m_bitangent);
+
+    OUT.m_tangent.w = IN.m_tangent.w;
+    return OUT;
+}
+
+struct PixelOutput
+{
+    float4 m_color : SV_Target0;
+};
+
+// Represent a vector in the (-1, -1, -1) to (1, 1, 1) range as a color in the (0, 0, 0) to (1, 1, 1) range
+// Color key
+// + x-axis: Light Coral
+// - x-axis: Teal
+// + y-axis: Bright Green
+// - y-axis: Dark Magenta
+// + z-axis: Medium Slate Blue
+// - z-axis: Olive
+float3 OffsetColor(float3 color)
+{
+    return normalize(color) * 0.5 + 0.5;
+}
+
+PixelOutput MainPS(VSOutput IN)
+{
+    PixelOutput OUT;
+    
+    float3 tangents[UvSetCount] = { IN.m_tangent.xyz, IN.m_tangent.xyz };
+    float3 bitangents[UvSetCount] = { IN.m_bitangent.xyz, IN.m_bitangent.xyz };
+    
+    if(o_reconstructionMethod == ReconstructionMethods::ReconstructBitangent)
+    {
+        bitangents[0] = cross(IN.m_normal.xyz, IN.m_tangent.xyz) * sign(IN.m_tangent.w);
+    }
+    else if(o_reconstructionMethod == ReconstructionMethods::UseSurfaceGradient)
+    {
+        const bool isBackface = false;
+        SurfaceGradientNormalMapping_Init(IN.m_normal, IN.m_worldPosition, isBackface);
+        for (uint i = 0; i < UvSetCount; ++i)
+        {
+            SurfaceGradientNormalMapping_GenerateTB(IN.m_uv[i], tangents[i], bitangents[i]); 
+        } 
+    }
+
+
+    float3 outColor = float3(1.0, 1.0, 1.0);
+    switch(o_debugVertexStream)
+        {
+        case DebugVertexStream::Normals:
+            outColor = OffsetColor(IN.m_normal);
+            break;
+        case DebugVertexStream::Tangents:
+            outColor = OffsetColor(tangents[0]);
+            break;
+        case DebugVertexStream::Bitangents:
+            outColor = OffsetColor(bitangents[0]);
+            break;
+        case DebugVertexStream::UVs:
+            // Assume a tiled uv visualization, where anything greater than 1 wraps back around to 0
+            outColor = float3(frac(IN.m_uv[0].x), frac(IN.m_uv[0].y), 0.0f);
+            break;            
+        case DebugVertexStream::TangentW:
+            float red = IN.m_tangent.w >= 0.0f ? 1.0f : 0.0f;
+            float green = IN.m_tangent.w < 0.0f ? 1.0f : 0.0f;
+            outColor = float3(red, green, 0.0f);
```

Comment:
I actually recommend using blue for a `w` component that's exactly 0 (which would indicate a baking error I believe)

Reply:
Is it possible for a single triangle to have vertices with both 1 and -1 for the tangent.w? I actually had that initially, but then I was thinking it might be a matter of interpolating between 1 and -1, which could land on 0.

I would hope that any vertex where the adjacent triangles used a different sign for tangent.w would be duplicated as two vertices with the same position but different tangent.w values, but I didn't know if that was actually the case. I know we're at least not welding vertices with different tangent.w values, but I don't know if modeling tools ensure that.

Reply:
I don't believe those vertices would be welded. Either way we'd probably want to know if it was happening I imagine

Reply:
I guess that would be the baking error you're referring to? If it's zero, that means some tool baked out the vertices improperly leading to a triangle with differing tangent.w values in its vertices.

Reply:
Correct

Reply:
Done.